File: leetcode/partitionLabels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        vector<int> last(26);
        for(int i = 0; i < s.size(); i++) last[s[i] - 'a'] = i;
        
        int l = -1;
        int r = 0;
        vector<int> ans;
        
        for(int i = 0; i < s.size(); i++) {
            r = (r > last[s[i] - 'a']) ? r : last[s[i] - 'a'];
            if(r == i) {
                ans.push_back(r - l);
                l = i;
            }
        }
        
        return ans;
    }
};

File: leetcode/partitionLabels.java
class Solution {

  public List<Integer> partitionLabels(String s) {
      int last[] = new int[26];
      for(int i = 0; i < s.length(); i++) last[s.charAt(i) - 'a'] = i;
      
      int l = -1;
      int r = 0;
      List<Integer> ans = new ArrayList<>();
      
      for(int i = 0; i < s.length(); i++) {
          r = (r > last[s.charAt(i) - 'a'])? r : last[s.charAt(i) - 'a'];
          if(i == r) {
              ans.add(r - l);
              l = i;
          }
      }
      
      return ans;
  }
}